### PR TITLE
Use map instead of each for ActiveRecord::Relation#destroy_all

### PR DESCRIPTION
**Pull request by glupo**
### Summary
As per documentation `destroy_all` should return a collection of destroyed records, but since it uses `each(&:destroy)` it instead returns a collection it attempted to destroy. Made it to match the current documentation but maybe it should have more informative response ( somehow notify about records that failed to be destroyed )

### CODE REVIEW

**Comment by kamipo** [comment]
Thanks, this is fine as is.

**Comment by rafaelfranca** [comment]
In that case we need to change the documentation that say that it sometimes will return some objects that we not destroyed for some reason.

**Comment by kamipo** [comment]
`Persistence#destroy` will return destroyed object or raise any exception, not return not destroyed object.

https://github.com/rails/rails/blob/604ef720671adf658da2b9c9c4af7ffa350c2162/activerecord/lib/active_record/persistence.rb#L325-L336

**Comment by rafaelfranca** [comment]
:+1:

**Comment by eugeneius** [comment]
From the documentation for that method:

https://github.com/rails/rails/blob/604ef720671adf658da2b9c9c4af7ffa350c2162/activerecord/lib/active_record/persistence.rb#L321-L323

`Persistence#destroy` is overridden by `Callbacks#destroy`: https://github.com/rails/rails/blob/604ef720671adf658da2b9c9c4af7ffa350c2162/activerecord/lib/active_record/callbacks.rb#L323

So it _is_ possible for `record.destroy` to fail and return `false`.

**Comment by kamipo** [comment]
hm, you are right. So if we want to change the current behavior, need a CHANGELOG entry since it is a breaking change.

File: activerecord/lib/active_record/relation.rb [hunk 1]
```diff
@@ -346,7 +346,7 @@ def update_all(updates)
     #
     #   Person.where(age: 0..18).destroy_all
     def destroy_all
-      records.each(&:destroy).tap { reset }
+      records.map(&:destroy).select{ |record| record.present? }.tap { reset }
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation.rb]: Don't we want `.select(&:destroyed?)`?

**Review by rafaelfranca — COMMENTED** [review]
It is also missing a test for this new behavior.

File: activerecord/lib/active_record/relation.rb (hunk 1 above)
Comment (sgrif) [reply activerecord/lib/active_record/relation.rb]: This can actually be shortened to just `select(&:destroy)`

**Comment by sgrif** [comment]
> need a CHANGELOG entry since it is a breaking change.

It's a "breaking change" in that it changes the return value... However, given that the return value differs from what is documented, I'd call that a bug fix.

**Comment by kamipo** [comment]
At least we have another `.each(&:destroy)`. Currently both `klass.destroy(ids)` and `relation.destroy_all` will return matched objects.
If we will change `relation.destroy_all` to return affected objects, I feel like that behavior will be inconsistently to the other.

https://github.com/rails/rails/blob/f8afb5168aed810a061d081d3941c9dc9bd633c3/activerecord/lib/active_record/persistence.rb#L138-L144

**Comment by sgrif** [comment]
`Klass.destroy(id)` doesn't document its return value. I agree this should remain consistent. I think it makes sense to change that method as well to be consistent with this one, and change the documentation to mention this.

It's also worth mentioning that we can do a deprecation cycle on this pretty easily. I don't think it's worth doing here, this is very clearly a bug, but I wanted to mention it.

**Comment by glupo** [comment]
@rafaelfranca I've added requested changes but I'm actually not sure where to test this. I feel like tests for it should take place in `relations_test.rb` but we need to add some kind of `before_destroy` callback and I'm not sure I should ad it to that file. Can you give me a little hint on how should I do it?

**Comment by eugeneius** [comment]
You could try using the `FailedBulb` model, which already has a `before_destroy` callback that halts the callback chain:
https://github.com/rails/rails/blob/0c743885baba1681ad43376f6444b52a306fdfd9/activerecord/test/models/bulb.rb#L50-L54

**Pull request closed without merging**